### PR TITLE
Turn on NSSOCSP check in mod_nss conf

### PR DESCRIPTION
**Pull request by pvomacka**
Turn on NSSOCSP directive during install/replica install/upgrade.
That check whether the certificate which is used for login is
revoked or not using OSCP.

https://pagure.io/freeipa/issue/6370

### CODE REVIEW

File: ipaserver/install/httpinstance.py [hunk 1]
```diff
@@ -156,6 +157,7 @@ def create_instance(self, realm, fqdn, domain_name, pkcs12_info=None,
                   self.set_mod_nss_protocol)
         self.step("setting mod_nss password file", self.__set_mod_nss_passwordfile)
         self.step("enabling mod_nss renegotiate", self.enable_mod_nss_renegotiate)
+        self.step("enabling mod_nss NSSOCSP", self.enable_mod_nss_ocsp)
```
Comment (rcritten) [inline ipaserver/install/httpinstance.py]: nit: I think OCSP is preferrable over NSSOCSP which may confuse some people.

File: ipaserver/install/httpinstance.py [hunk 2]
```diff
@@ -263,6 +265,19 @@ def enable_mod_nss_renegotiate(self):
...
+        ocsp_path = os.path.join('/files',
+                                 paths.HTTPD_NSS_CONF[1:],
+                                 'VirtualHost')
+        ocsp = aug.get(os.path.join(ocsp_path, 'directive[. = "NSSOCSP"]/arg'))
+
+        if ocsp is None:
+            aug.set(os.path.join(ocsp_path, 'directive[last()+1]'), 'NSSOCSP')
+
+        aug.set(os.path.join(ocsp_path, 'directive[. = "NSSOCSP"]/arg'), 'on')
+        aug.save()
```
Comment (rcritten) [inline ipaserver/install/httpinstance.py]: I pulled this into a separate file and it didn't seem to work in F-25, could be something I did wrong...

File: ipaserver/install/server/upgrade.py [hunk 3]
```diff
@@ -1375,7 +1375,7 @@ def remove_ds_ra_cert(subject_base):
 def fix_trust_flags():
     root_logger.info('[Fixing trust flags in %s]' % paths.HTTPD_ALIAS_DIR)
 
-    if sysupgrade.get_upgrade_state('http', 'fix_trust_flags'):
+    if sysupgrade.get_upgrade_state('http', 'fix_trust_flags2'):
```
Comment (rcritten) [inline ipaserver/install/server/upgrade.py]: I don't understand why you are changing the name of the state rather than adding a new state.

File: ipaserver/install/server/upgrade.py [hunk 4]
```diff
@@ -1389,7 +1389,11 @@ def fix_trust_flags():
     if cert:
         db.trust_root_cert(nickname, 'CT,C,C')
 
-    sysupgrade.set_upgrade_state('http', 'fix_trust_flags', True)
+    sc_nickname = installutils.get_directive(paths.HTTPD_NSS_CONF,
+                                             "NSSNickname")
```
Comment (rcritten) [inline ipaserver/install/server/upgrade.py]: Should there be a None check here for paranoia reasons?

File: ipaserver/install/server/upgrade.py [hunk 5]
```diff
@@ -1404,6 +1408,11 @@ def update_mod_nss_protocol(http):
     sysupgrade.set_upgrade_state('nss.conf', 'protocol_updated_tls12', True)
 
 
+def enable_mod_nss_ocsp(http):
+    root_logger.info('[Updating mod_nss enabling NSSOCSP]')
```
Comment (rcritten) [inline ipaserver/install/server/upgrade.py]: Similarly, I'd replace NSSOCSP with just OCSP.

File: ipaserver/install/httpinstance.py (hunk 1 above)
Comment (pvomacka) [inline ipaserver/install/httpinstance.py]: Thats true, I'll change that.

File: ipaserver/install/httpinstance.py (hunk 2 above)
Comment (pvomacka) [inline ipaserver/install/httpinstance.py]: I tried it during IPA installation, not alone and it worked.

File: ipaserver/install/server/upgrade.py (hunk 3 above)
Comment (pvomacka) [inline ipaserver/install/server/upgrade.py]: Yes, I could create new state, but is it really needed? My new code fixes trust flags in httpd/alias, thats what this state is for, so I don't see any point in copying (duplicating) it.

File: ipaserver/install/server/upgrade.py (hunk 4 above)
Comment (pvomacka) [inline ipaserver/install/server/upgrade.py]: I'm not sure whether there can be a situation when NSSNickname is not specified. Is it possible or not?

File: install/restart_scripts/restart_httpd [hunk 6]
```diff
@@ -21,11 +21,25 @@
...
 from ipaserver.install import certs
 
 
 def _main():
+
+    api.bootstrap(in_server=True, context='restart', confdir=paths.ETC_IPA)
+    api.finalize()
+
+    db = certs.CertDB(api.env.realm, nssdir=paths.HTTPD_ALIAS_DIR)
+    server_certs = db.find_server_certs()
```
Comment (flo-renaud) [inline install/restart_scripts/restart_httpd]: db.find_server_certs() returns all the certs in the DB with the u flag, and it may return more than one nickname, maybe not the one you are expecting. You could use installutils.get_directive(paths.HTTPD_NSS_CONF,'NSSNickname') instead in order to find the nickname of HTTPd server certificate.

File: ipaserver/install/server/upgrade.py (hunk 3 above)
Comment (stlaz) [inline ipaserver/install/server/upgrade.py]: Rob is right that you should probably have a new state to do this, this one was supposed to do some changes and you're adding some more on top of that.

File: install/restart_scripts/restart_httpd [hunk 7]
```diff
@@ -21,11 +21,25 @@
...
+    if len(server_certs) == 0:
+        raise RuntimeError("Could not find a suitable server cert.")
+
+    nickname = server_certs[0][0]
+    db.trust_root_cert(nickname, "P,,")
+
```
Comment (flo-renaud) [inline install/restart_scripts/restart_httpd]: Is there a specific reason for changing the flag in this script (during restart of httpd) rather than during upgrade? Is it to make sure that certs configured with ipa-server-certinstall have the proper flag? Then a comment would help future developers.

File: ipaserver/install/server/upgrade.py (hunk 3 above)
Comment (pvomacka) [inline ipaserver/install/server/upgrade.py]: Ok, at the beginning I wanted to add it, too,  but this solution (not add new state) was recommended me by @HonzaCholasta. So if @HonzaCholasta is ok with a new state, then I will create it.

File: ipaserver/install/server/upgrade.py (hunk 4 above)
Comment (rcritten) [inline ipaserver/install/server/upgrade.py]: Not sure but throughout the cert code whenever a cert is searched for in this way a sanity check follows.

File: install/restart_scripts/restart_httpd (hunk 7 above)
Comment (rcritten) [inline install/restart_scripts/restart_httpd]: This is the certmonger restart script. This is probably a belt-and-suspenders change to assure that the peer flag is retained. I agree a comment would be helpful.

File: ipaserver/install/httpinstance.py (hunk 2 above)
Comment (rcritten) [inline ipaserver/install/httpinstance.py]: Ok, I'll chalk it up to a copy-paste problem on my end assuming others do an end-to-end test and confirm all three possible states (No starting NSSOCSP, NSSOCSP off, and NSSOCSP already on.

File: install/restart_scripts/restart_httpd (hunk 6 above)
Comment (pvomacka) [inline install/restart_scripts/restart_httpd]: Hello @flo-renaud, thank you for review. 
Yes, I that will be better to use the installutils method.

File: install/restart_scripts/restart_httpd (hunk 7 above)
Comment (pvomacka) [inline install/restart_scripts/restart_httpd]: Yes, @rcritten is right. And true is that I forgot to add it into commit message. I will add it there.

File: install/restart_scripts/restart_httpd [hunk 8]
```diff
@@ -21,11 +21,22 @@
...
 def _main():
+
+    api.bootstrap(in_server=True, context='restart', confdir=paths.ETC_IPA)
+    api.finalize()
+
+    db = certs.CertDB(api.env.realm, nssdir=paths.HTTPD_ALIAS_DIR)
+    nickname = installutils.get_directive(paths.HTTPD_NSS_CONF, "NSSNickname")
+
+    db.trust_root_cert(nickname, "P,,")
+
```
Comment (rcritten) [inline install/restart_scripts/restart_httpd]: I think a one-line comment here is needed so future devs will know what this is about without having to dig thru git history.

File: ipaserver/install/server/upgrade.py [hunk 9]
```diff
@@ -1375,6 +1375,11 @@ def remove_ds_ra_cert(subject_base):
 def fix_trust_flags():
     root_logger.info('[Fixing trust flags in %s]' % paths.HTTPD_ALIAS_DIR)
 
+    db = certs.CertDB(api.env.realm, nssdir=paths.HTTPD_ALIAS_DIR)
+    sc_nickname = installutils.get_directive(paths.HTTPD_NSS_CONF,
+                                             "NSSNickname")
+    db.trust_root_cert(sc_nickname, "P,,")
+
```
Comment (rcritten) [inline ipaserver/install/server/upgrade.py]: This will ALWAYS fix the trust though. Is that what you want? I think it should be in a separate function, tracked separately, with a similar test for being already executed. I doubt it would cause any harm being run multiple times but it is bad form.

File: ipaserver/install/httpinstance.py [hunk 10]
```diff
@@ -263,6 +265,19 @@ def enable_mod_nss_renegotiate(self):
...
+        aug = augeas.Augeas()
+        ocsp_path = os.path.join('/files',
+                                 paths.HTTPD_NSS_CONF[1:],
+                                 'VirtualHost')
+        ocsp = aug.get(os.path.join(ocsp_path, 'directive[. = "NSSOCSP"]/arg'))
+
+        if ocsp is None:
+            aug.set(os.path.join(ocsp_path, 'directive[last()+1]'), 'NSSOCSP')
+
+        aug.set(os.path.join(ocsp_path, 'directive[. = "NSSOCSP"]/arg'), 'on')
```
Comment (HonzaCholasta) [inline ipaserver/install/httpinstance.py]: You can shorten this to:
```python
        aug.set('/files/etc/httpd/conf.d/nss.conf/VirtualHost/directive[. = "NSSOCSP"]', 'NSSOCSP')
        aug.set('/files/etc/httpd/conf.d/nss.conf/VirtualHost/directive[. = "NSSOCSP"]/arg', 'on')
```
Additionally, you can prepend the following to uncomment NSSOCSP:
```python
        aug.set('/files/etc/httpd/conf.d/nss.conf/VirtualHost/#comment[. =~ regexp("NSSOCSP .*")]', 'NSSOCSP')
        aug.rename('/files/etc/httpd/conf.d/nss.conf/VirtualHost/#comment[. = "NSSOCSP"]', 'directive')
```

**Comment by HonzaCholasta** [comment]
@pvomacka, CI fails because you forgot to include python-augeas in lint BuildRequires.

File: ipaserver/install/server/upgrade.py [hunk 11]
```diff
@@ -1392,6 +1392,23 @@ def fix_trust_flags():
     sysupgrade.set_upgrade_state('http', 'fix_trust_flags', True)
 
 
+def fix_server_cert_trust_flags():
```
Comment (HonzaCholasta) [inline ipaserver/install/server/upgrade.py]: Nitpick, but I would rather name this `http_trust_server_cert` or similar.

File: ipaserver/install/httpinstance.py [hunk 12]
```diff
@@ -263,6 +265,20 @@ def enable_mod_nss_renegotiate(self):
         installutils.set_directive(paths.HTTPD_NSS_CONF, 'NSSRenegotiation', 'on', False)
         installutils.set_directive(paths.HTTPD_NSS_CONF, 'NSSRequireSafeNegotiation', 'on', False)
 
+    def enable_mod_nss_ocsp(self):
+        aug = augeas.Augeas()
+        ocsp_path = os.path.join('/files',
+                                 paths.HTTPD_NSS_CONF[1:],
+                                 'VirtualHost')
```
Comment (HonzaCholasta) [inline ipaserver/install/httpinstance.py]: You should use format strings here as well:
```python
        ocsp_path = '/files{}/VirtualHost'.format(paths.HTTPD_NSS_CONF)
```

File: ipaserver/install/httpinstance.py [hunk 13]
```diff
@@ -263,6 +265,20 @@ def enable_mod_nss_renegotiate(self):
...
+        aug.set(
+            '{path}/#comment[. =~ regexp("NSSOCSP .*")]'.format(path=ocsp_path),
+            'NSSOCSP')
+        aug.rename('{path}/#comment[. = "NSSOCSP"]'.format(path=ocsp_path),
+                   'directive')
+        aug.set('{path}/directive[. = "NSSOCSP"]/arg'.format(path=ocsp_path),
+                'on')
```
Comment (HonzaCholasta) [inline ipaserver/install/httpinstance.py]: Nitpick: You can fit each of these statements on a single line by removing some unnecessary stuff:
```python
        aug.set('{}/#comment[.=~regexp("NSSOCSP .*")]'.format(path), 'NSSOCSP')
        aug.rename('{}/#comment[.="NSSOCSP"]'.format(path), 'directive')
        aug.set('{}/directive[.="NSSOCSP"]/arg'.format(path), 'on')
```

File: ipaserver/install/httpinstance.py [hunk 14]
```diff
@@ -263,6 +265,20 @@ def enable_mod_nss_renegotiate(self):
         installutils.set_directive(paths.HTTPD_NSS_CONF, 'NSSRenegotiation', 'on', False)
         installutils.set_directive(paths.HTTPD_NSS_CONF, 'NSSRequireSafeNegotiation', 'on', False)
 
+    def enable_mod_nss_ocsp(self):
+        aug = augeas.Augeas()
```
Comment (HonzaCholasta) [inline ipaserver/install/httpinstance.py]: By default, augeas loads *all* lenses and *all* configuration files it finds in the system. I guess we don't want it to do that and load only the httpd lens and the /etc/httpd/conf.d/nss.conf file, which can be done using:
```python
        aug = Augeas(flags=Augeas.NO_LOAD | Augeas.NO_MODL_AUTOLOAD)

        aug.set('/augeas/load/Httpd/lens', 'Httpd.lens')
        aug.set('/augeas/load/Httpd/incl', paths.HTTPD_NSS_CONF)
        aug.load()
```

File: freeipa.spec.in [hunk 15]
```diff
@@ -306,6 +306,7 @@ Requires: oddjob
 Requires: gssproxy >= 0.7.0-2
 # 1.15.2: FindByNameAndCertificate (https://pagure.io/SSSD/sssd/issue/3050)
 Requires: sssd-dbus >= 1.15.2
+Requires: python-augeas
```
Comment (tiran) [inline freeipa.spec.in]: Do we really want to depend on a new Python library, C library and complex framework just to change a line in a config file? I'm -1. The task can be accomplished with a bit of Python code and regexp.

Comment (pvomacka) [inline freeipa.spec.in]: Yes, it would be possible to do it only with python regexp but it will be much nicer to do it using augeas. And python-augeas is available in RHEL and Fedora so it is not a big issue to add a dependency.

Comment (HonzaCholasta) [inline freeipa.spec.in]: We want to migrate *all* config file handling to augeas eventually, now is as good time as any to add the dependency.

**Comment by flo-renaud** [comment]
Hi @pvomacka 
I tested your last update with a new install and with an upgraded instance, and both are functionally OK. Revoked certs do not allow to access IPA Web UI.

**Comment by MartinBasti** [comment]
augeas should be dependency of python2-ipaserver and python3-ipaserver (python3-augeas) packages
```
************* Module ipaserver.install.httpinstance
ipaserver/install/httpinstance.py:32: [E0401(import-error), ] Unable to import 'augeas')
```

**Comment by pvomacka** [comment]
Hello @flo-renaud, thank you for testing this. 
Hello @MartinBasti, thank you for review. I just fixed that.

**Comment by MartinBasti** [comment]
And you also need python[3]-augeas as Pylint BuildDependency to pass pylint :), sorry I forgot about it.

**Comment by MartinBasti** [comment]
And you also need to add it in `ipaserver/setup.py` as dependency for our PyPI packages

**Comment by pvomacka** [comment]
@MartinBasti thank you for comments, fixed.

**Comment by tomaskrizek** [comment]
master:

* e0b32dac5462164869ab19c3d56c36e80cde4b7b Turn on NSSOCSP check in mod_nss conf


ipa-4-5:

* 4aa7e70fcd1851394f943da669d6af4e11b60940 Turn on NSSOCSP check in mod_nss conf
